ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp

```diff
@@ -0,0 +1,392 @@
+
+	int bVelSource = 1;
+	nh.getParam("/op_global_planner/velocitySource", bVelSource);
+	if(bVelSource == 0)
+		sub_robot_odom 			= nh.subscribe("/odom", 					100,	&GlobalPlanner::callbackGetRobotOdom, 	this);
+	else if(bVelSource == 1)
+		sub_current_velocity 	= nh.subscribe("/current_velocity",		100,	&GlobalPlanner::callbackGetVehicleStatus, 	this);
+	else if(bVelSource == 2)
+		sub_can_info 			= nh.subscribe("/can_info",		100,	&GlobalPlanner::callbackGetCanInfo, 	this);
+
+}
+
+GlobalPlanner::~GlobalPlanner()
+{
+	if(m_params.bEnableRvizInput)
+		SaveSimulationData();
+}
+
+void GlobalPlanner::callbackGetGoalPose(const geometry_msgs::PoseStampedConstPtr &msg)
+{
+	PlannerHNS::WayPoint wp = PlannerHNS::WayPoint(msg->pose.position.x+m_OriginPos.position.x, msg->pose.position.y+m_OriginPos.position.y, msg->pose.position.z+m_OriginPos.position.z, tf::getYaw(msg->pose.orientation));
+	m_GoalsPos.push_back(wp);
+	ROS_INFO("Received Goal Pose");
+}
+
+void GlobalPlanner::callbackGetStartPose(const geometry_msgs::PoseWithCovarianceStampedConstPtr &msg)
+{
+	m_CurrentPose = PlannerHNS::WayPoint(msg->pose.pose.position.x+m_OriginPos.position.x, msg->pose.pose.position.y+m_OriginPos.position.y, msg->pose.pose.position.z+m_OriginPos.position.z, tf::getYaw(msg->pose.pose.orientation));
+	ROS_INFO("Received Start pose");
+}
+
+void GlobalPlanner::callbackGetCurrentPose(const geometry_msgs::PoseStampedConstPtr& msg)
+{
+	m_CurrentPose = PlannerHNS::WayPoint(msg->pose.position.x, msg->pose.position.y, msg->pose.position.z, tf::getYaw(msg->pose.orientation));
+}
+
+void GlobalPlanner::callbackGetRobotOdom(const nav_msgs::OdometryConstPtr& msg)
+{
+	m_VehicleState.speed = msg->twist.twist.linear.x;
+	m_CurrentPose.v = m_VehicleState.speed;
+	m_VehicleState.steer += atan(2.7 * msg->twist.twist.angular.z/msg->twist.twist.linear.x);
+}
+
+void GlobalPlanner::callbackGetVehicleStatus(const geometry_msgs::TwistStampedConstPtr& msg)
+{
+	m_VehicleState.speed = msg->twist.linear.x;
+	m_CurrentPose.v = m_VehicleState.speed;
+	if(fabs(msg->twist.linear.x) > 0.25)
+		m_VehicleState.steer = atan(2.7 * msg->twist.angular.z/msg->twist.linear.x);
+	UtilityHNS::UtilityH::GetTickCount(m_VehicleState.tStamp);
+}
+
+void GlobalPlanner::callbackGetCanInfo(const autoware_msgs::CanInfoConstPtr &msg)
+{
+	m_VehicleState.speed = msg->speed/3.6;
+	m_CurrentPose.v = m_VehicleState.speed;
+	m_VehicleState.steer = msg->angle * 0.45 / 660;
+	UtilityHNS::UtilityH::GetTickCount(m_VehicleState.tStamp);
+}
+
+bool GlobalPlanner::GenerateGlobalPlan(PlannerHNS::WayPoint& startPoint, PlannerHNS::WayPoint& goalPoint, std::vector<std::vector<PlannerHNS::WayPoint> >& generatedTotalPaths)
+{
+	std::vector<int> predefinedLanesIds;
+	double ret = 0;
+
+	ret = m_PlannerH.PlanUsingDP(startPoint, goalPoint, MAX_GLOBAL_PLAN_DISTANCE, m_params.bEnableLaneChange, predefinedLanesIds, m_Map, generatedTotalPaths);
+
+	if(ret == 0)
```

Comment:
ret is double, you can not compare for equality. Use:
```
// bool cmpf(float A, float B, float epsilon = 0.005f)
      //{
      //return (fabs(A - B) < epsilon);
      //}
```

Reply:
here 0 means no path is generated and it is dead zero by design.

Reply:
I think you did not understand my point. you can not compare floating point numbers, please see: http://floating-point-gui.de/errors/comparison/.